File: crypto/src/main/java/io/oversec/one/crypto/BaseDecryptResult.java
package io.oversec.one.crypto;


import com.google.protobuf.InvalidProtocolBufferException;
import io.oversec.one.crypto.proto.Inner;

import java.io.UnsupportedEncodingException;

public class BaseDecryptResult {


    public enum Result {
        OK, RETRY, FAILED_PERMANENTLY
    }

    public enum DecryptError {
        SYM_UNSUPPORTED_CIPHER, SYM_NO_MATCHING_KEY, SYM_DECRYPT_FAILED, PGP_ERROR, PROTO_ERROR, NO_HANDLER
    }


    protected EncryptionMethod mEncryptionMethod;


    protected byte[] mDecryptedRawData;
    protected Result mResult;
    protected DecryptError mError;

    protected String mErrorMessage;


    public BaseDecryptResult(EncryptionMethod encryptionMethod, byte[] rawData) {
        mEncryptionMethod = encryptionMethod;
        mDecryptedRawData = rawData;
        mResult = Result.OK;
    }

    public BaseDecryptResult(EncryptionMethod method, DecryptError error) {
        this(method, error, null);
    }


    public BaseDecryptResult(EncryptionMethod method, DecryptError error, String errorMessage) {
        mEncryptionMethod = method;
        mError = error;
        mErrorMessage = errorMessage;
        mResult = Result.RETRY;
        switch (error) {
            case SYM_DECRYPT_FAILED:
                mResult = Result.RETRY;
                break;
            case SYM_NO_MATCHING_KEY:
                mResult = Result.RETRY;
                break;
            case PGP_ERROR:
                new Exception().printStackTrace();
                mResult = Result.RETRY; //TODO: should be failed permanently or?
                break;
            case PROTO_ERROR:
                mResult = Result.FAILED_PERMANENTLY;
                break;
            case NO_HANDLER:
                mResult = Result.RETRY;
                break;
            case SYM_UNSUPPORTED_CIPHER:
                mResult = Result.RETRY;
                break;

        }
    }


    public boolean isOk() {
        return mResult == Result.OK;
    }

    public boolean isRetry() {
        return mResult == Result.RETRY;
    }

    public boolean isOversecNode() {
        return mResult == Result.OK
                || mError == DecryptError.SYM_NO_MATCHING_KEY
                || mError == DecryptError.SYM_UNSUPPORTED_CIPHER
                || mError == DecryptError.SYM_DECRYPT_FAILED
                || mError == DecryptError.NO_HANDLER
                || mError == DecryptError.PGP_ERROR;
    }


    public EncryptionMethod getEncryptionMethod() {
        return mEncryptionMethod;
    }


    public String getErrorMessage() {
        return mErrorMessage;
    }


    public Inner.InnerData getDecryptedDataAsInnerData() throws InvalidProtocolBufferException {
        return Inner.InnerData.parseFrom(mDecryptedRawData);
    }
    public String getDecryptedDataAsUtf8String() throws UnsupportedEncodingException {
        return new String(mDecryptedRawData,"UTF-8");
    }

    public Result getResult() {
        return mResult;
    }

    public DecryptError getError() {
        return mError;
    }

    @Override
    public String toString() {
        return "BaseDecryptResult{" +
                "decryptedData='" + mDecryptedRawData + '\'' +
                ", result=" + mResult +
                ", error=" + mError +
                '}';
    }


}

File: crypto/src/main/java/io/oversec/one/crypto/BaseDecryptResult.kt
package io.oversec.one.crypto

import com.google.protobuf.InvalidProtocolBufferException
import io.oversec.one.crypto.proto.Inner
import java.io.UnsupportedEncodingException
import java.nio.charset.Charset

open class BaseDecryptResult {

    enum class Result {
        OK, RETRY, FAILED_PERMANENTLY
    }

    enum class DecryptError {
        SYM_UNSUPPORTED_CIPHER, SYM_NO_MATCHING_KEY, SYM_DECRYPT_FAILED, PGP_ERROR, PROTO_ERROR, NO_HANDLER
    }

    open var encryptionMethod: EncryptionMethod?
        protected set

    protected var mDecryptedRawData: ByteArray? = null

    var result: Result
        protected set

    var error: DecryptError? = null
        protected set

    var errorMessage: String? = null
        protected set

    constructor(method: EncryptionMethod, rawData: ByteArray) {
        encryptionMethod = method
        mDecryptedRawData = rawData
        result = Result.OK
    }

    @JvmOverloads
    constructor(method: EncryptionMethod?, err: DecryptError, errMsg: String? = null) {
        encryptionMethod = method
        error = err
        errorMessage = errMsg
        result = Result.RETRY
        when (error) {
            BaseDecryptResult.DecryptError.SYM_DECRYPT_FAILED -> result = Result.RETRY
            BaseDecryptResult.DecryptError.SYM_NO_MATCHING_KEY -> result = Result.RETRY
            BaseDecryptResult.DecryptError.PGP_ERROR -> {
                Exception().printStackTrace()
                result = Result.RETRY //TODO: should be failed permanently or?
            }
            BaseDecryptResult.DecryptError.PROTO_ERROR -> result = Result.FAILED_PERMANENTLY
            BaseDecryptResult.DecryptError.NO_HANDLER -> result = Result.RETRY
            BaseDecryptResult.DecryptError.SYM_UNSUPPORTED_CIPHER -> result = Result.RETRY
        }
    }

    val isOk: Boolean
        get() = result == Result.OK

    val isRetry: Boolean
        get() = result == Result.RETRY

    val isOversecNode: Boolean
        get() = (result == Result.OK
                || error == DecryptError.SYM_NO_MATCHING_KEY
                || error == DecryptError.SYM_UNSUPPORTED_CIPHER
                || error == DecryptError.SYM_DECRYPT_FAILED
                || error == DecryptError.NO_HANDLER
                || error == DecryptError.PGP_ERROR)


    val decryptedDataAsInnerData: Inner.InnerData
        @Throws(InvalidProtocolBufferException::class)
        get() = Inner.InnerData.parseFrom(mDecryptedRawData)

    val decryptedDataAsUtf8String: String
        @Throws(UnsupportedEncodingException::class)
        get() = String(requireNotNull(mDecryptedRawData), Charset.forName("UTF-8"))

    override fun toString(): String {
        return "BaseDecryptResult{" +
                "decryptedData='" + mDecryptedRawData + '\''.toString() +
                ", result=" + result +
                ", error=" + error +
                '}'.toString()
    }
}
